qiskit/algorithms/evolvers/pvqd/pvqd.py

```diff
@@ -0,0 +1,390 @@
+                timestep=0.01,
+                optimizer=optimizer,
+                quantum_instance=backend,
+                expectation=expectation
+            )
+
+            # specify the evolution problem
+            problem = EvolutionProblem(
+                hamiltonian, time, aux_operators=[hamiltonian, observable]
+            )
+
+            # and evolve!
+            result = pvqd.evolve(problem)
+
+    References:
+
+        [1] Stefano Barison, Filippo Vicentini, and Giuseppe Carleo (2021), An efficient
+            quantum algorithm for the time evolution of parameterized circuits,
+            `Quantum 5, 512 <https://quantum-journal.org/papers/q-2021-07-28-512/>`_.
+    """
+
+    def __init__(
+        self,
+        ansatz: Optional[QuantumCircuit] = None,
+        initial_parameters: Optional[np.ndarray] = None,
+        timestep: Optional[float] = None,
+        optimizer: Optional[Union[Optimizer, Minimizer]] = None,
+        expectation: Optional[ExpectationBase] = None,
+        initial_guess: Optional[np.ndarray] = None,
+        evolution: Optional[EvolutionSynthesis] = None,
+        use_parameter_shift: bool = True,
+        quantum_instance: Optional[Union[Backend, QuantumInstance]] = None,
+    ) -> None:
+        """
+        Args:
+            ansatz: A parameterized circuit preparing the variational ansatz to model the
+                time evolved quantum state.
+            initial_parameters: The initial parameters for the ansatz. Together with the ansatz,
+                these define the initial state of the time evolution.
+            timestep: The time step.
+            optimizer: The classical optimizers used to minimize the overlap between
+                Trotterization and ansatz. Can be either a :class:`.Optimizer` or a callable
+                using the :class:`.Minimizer` protocol.
+            expectation: The expectation converter to evaluate expectation values.
+            initial_guess: The initial guess for the first VQE optimization. Afterwards the
+                previous iteration result is used as initial guess. If None, this is set to
+                a random vector with elements in the interval :math:`[-0.01, 0.01]`.
+            evolution: The evolution synthesis to use for the construction of the Trotter step.
+                Defaults to first-order Lie-Trotter decomposition, see also
+                :mod:`~qiskit.synthesis.evolution` for different options.
+            use_parameter_shift: If True, use the parameter shift rule to compute gradients.
+                If False, the optimizer will not be passed a gradient callable. In that case,
+                Qiskit optimizers will use a finite difference rule to approximate the gradients.
+            quantum_instance: The backend of quantum instance used to evaluate the circuits.
+        """
+        if evolution is None:
+            evolution = LieTrotter()
+
+        self.ansatz = ansatz
```

Comment:
> And with None we have no default value so expect a user to set it later. With all Nones here this creates an object in arguably an invalid state where it can't be used.

That's why there's the `_validate_setup` check before the algorithms main method starts 🙂 This checks that all required arguments have been set.

Reply:
> And for timestep, where its a float - is there no float default value that is suitable, to have a valid value, instead of having it Optional and defaulting to None?

Since the accuracy of the algorithm depends a lot on the timestep I had it as required before. But I can set it to something like 0.01 that should be fine in most cases 🙂

Reply:
I get that the validate_setup checks later -  but I am more wondering why we don't do what is more normal and have the parameters we require passed in the constructor as required. Yes they can be updated if wanted later - but not be supplied (i.e. Optional) and have to be updated later that is what I a missing. The only time we normally do something like this is if the user does not know upfront, but we want to create the algorithm and pass it to something else to use. We did that with QusntumInstance to allow the backend to be set later. Now for a user code of an evolver it can for instance build out an EvolutionProblem and evolve it with this, well it can't if its not fully configured and that means it has to specifically understand PVQD to configure it right. Yes the user can do it all upfront but they are not obliged, and if they fail to configure it correctly it fails later during their computation when it gets to this point.

Reply:
Resolved by requiring required args 🙂